### PR TITLE
[Runtime][Bugfix] Added type-checking for Array::insert

### PR DESCRIPTION
**Pull request by Lunderberg**
Prior to this commit, the following code would compile and run without error.  This occurs because the typed `Array<T>::insert` calls the untyped `ArrayNode::InitRange`, with no type-checking done before the call.

```c++
Var x("x");
Var y("y");
Array<Var> var_arr{x, y};
Array<PrimExpr> expr_arr{x + 1, y + 2};
// Erroneously inserts static-type PrimExpr, runtime-type Add, even
// though neither PrimExpr is a type of Var.
var_arr.insert(var_arr.begin(), expr_arr.begin(), expr_arr.end());
```

After this commit, a `static_assert` in `Array<T>::insert` and in `Array<T>::Array(IterType,IterTYpe)` restricts the iterators, such that they must dereference to `T`, `Optional<T>`, a subclass of `T`, or `Optional<U>` where `U` is a subclass of `T`.

The public method `ArrayNode::SetItem` exposes a similar issue.  In the future, we may want to make it be private, accessed only through type-safe method in `Array<T>::Set`.

cc @areusch

### CODE REVIEW

**Comment by Lunderberg** [comment]
Breakage is in the WASM build, as the EMCC flags were still using `-std=c++14`.  Fixed in https://github.com/apache/tvm/pull/12693, will re-run CI after it lands.

**Comment by Lunderberg** [comment]
Breakage is in the iOS build, as the clang flags were still using `-std=gnu++14`.  Fixed in https://github.com/apache/tvm/pull/12712, will re-run CI after it lands.

**Comment by areusch** [comment]
we probably should resolve https://discuss.tvm.apache.org/t/downgrade-tvm-runtime-to-c-11/13492/5 before merging this

**Comment by Lunderberg** [comment]
From the discussion, it sounds like C++17 shouldn't be a blocker anymore, since there are no specific use cases at the moment, and multiple options to resolve it if/when they occur.

**Comment by Lunderberg** [comment]
Rebased onto main, as https://github.com/apache/tvm/pull/12692 also needed and implemented the same type-checking structs, which makes this PR much smaller.

File: src/te/schedule/schedule_lang.cc [hunk 1]
```diff
@@ -200,7 +200,7 @@ Stage& Stage::env_threads(Array<IterVar> threads) {
   ICHECK_EQ(self->env_threads.size(), 0U) << "Already set env_threads";
   Array<IterVar>& leaf_vars = self->leaf_iter_vars;
   Array<IterVar>& all_vars = self->all_iter_vars;
-  std::vector<ObjectRef> temp;
+  std::vector<IterVar> temp;
```
Comment (areusch) [inline src/te/schedule/schedule_lang.cc:203]: this change seems unrelated?

Comment (Lunderberg) [reply src/te/schedule/schedule_lang.cc:203]: This was actually the exact type of error that I wanted to catch with the `static_assert` statements.  This was a `std::vector<ObjectRef>` which was copied into an `Array<IterVar>`, with no type-checking during the conversion.  It didn't cause any runtime issues, because the `std::vector` was originally populated with `ObjectRef` instances that contain `IterVarNode`, but there were no compile-time checks to ensure that was the case.  Had it been otherwise, the later use of `ObjectRef::DowncastNoCheck<IterVar>` would have invoked undefined behavior.

Comment (areusch) [reply src/te/schedule/schedule_lang.cc:203]: oh i see, the diff hid line 207 by default, but now i see that temp gets placed into leaf_vars. that makes much more sense :)

**Review by areusch — APPROVED** [review]
thanks @Lunderberg !
